### ISSUE TITLE
Crash on Azure headless server with message "killed" in CLI

### ISSUE THREAD

**Issue by simon-winter** [Unity-Technologies/ml-agents#4731]
**Describe the bug**
I've setup an Agent with a few manual observations and a custom simple grid sensor that observes a 100x100 pixel minimap (https://github.com/mbaske/grid-sensor/tree/master/GridSensor).

This setup trains slow but fine on my pc so i outsourced it into an Azure VM. it trains also there for some time, but it cancels itself before writing any summary (25000 steps). it simply puts "killed" into the CLI after some time. i suspect it happens while writing summary, but ctrl+c before that will properly abort training and save me a model. incase i don't manually abort i just get "killed" and no model or more infos. i cannot resume training afterwards probably because something is still running, a restart of the vm fixes the issue till the next crash.

CLI output for training, failing and then trying to restart training. Note i tried to train in headless mode and also build the env as server build.

![grafik](https://user-images.githubusercontent.com/34577718/101680781-7db22e00-3a61-11eb-9c16-d3ca96788a09.png)
![grafik](https://user-images.githubusercontent.com/34577718/101680838-915d9480-3a61-11eb-8411-5ea70b42968d.png)
![grafik](https://user-images.githubusercontent.com/34577718/101680864-9a4e6600-3a61-11eb-8a5b-e9069b63936f.png)
![grafik](https://user-images.githubusercontent.com/34577718/101680938-b18d5380-3a61-11eb-84fb-f30e2ac758cd.png)

**Environment (please complete the following information):**
- Unity Version: Unity 2020.1.0b4
- OS + version: Linux server (ubuntu 18.04)
- _ML-Agents version_: release 10
- _Torch version_:1.7.0+cu92

I know thats not really much info nor reproducable, but maybe somebody has an idea? could it be the writing permission for the summary or something? Can i just install pytorch and mlagents on any ubuntu server and train there instead of using this weird Azure service?

**Comment by awjuliani** [comment]
Hi @simon-winter 

We currently do not actively support training on Azure. You are free to run mlagents on any linux cloud hosting service, as long as you have the correct permissions to install the necessary packages. Ubuntu should work fine.

**Comment by simon-winter** [comment]
i setup the training on a diffrent server hoster. same ubuntu and jsut installed the packages. i trained always with "--no-graphics and a headless build, nonetheless i get this run log:

![grafik](https://user-images.githubusercontent.com/34577718/101760656-70398a00-3adb-11eb-9f8d-25bb8121be92.png)

i will update this threat in a few hours if the current instance trains or crashes aswell

**Comment by simon-winter** [comment]
The new server does not cause my application to "kill" itself, atleast fo the time being. But i noticed that the training steps become subsequently slower, as in the first 10k steps pass in minutes while the next 10k need hours. This also happens on my local machine build, so i think the problem lys within the training or the environment itself and the former server just handled a possible memleak or similar differently. 
 this is the timings file of said behavior:

```
{
    "name": "root",
    "metadata": {
        "timer_format_version": "0.1.0",
        "start_time_seconds": "1607595274",
        "python_version": "3.6.9 (default, Oct  8 2020, 12:12:24) \n[GCC 8.4.0]",
        "command_line_arguments": "/usr/local/bin/mlagents-learn /root/explorerSAC_config.yaml --env=/root/BuildLinux/lin --run-id=sac --no-graphics --force",
        "mlagents_version": "0.22.0",
        "mlagents_envs_version": "0.22.0",
        "communication_protocol_version": "1.2.0",
        "pytorch_version": "1.7.0+cu92",
        "numpy_version": "1.19.4",
        "end_time_seconds": "1607595286"
    },
    "total": 12.59061484100016,
    "count": 1,
    "self": 0.033602250000285494,
    "children": {
        "run_training.setup": {
            "total": 0.003998755999873538,
            "count": 1,
            "self": 0.003998755999873538
        },
        "TrainerController.start_learning": {
            "total": 12.553013835000002,
            "count": 1,
            "self": 0.03308288199968956,
            "children": {
                "TrainerController._reset_env": {
                    "total": 0.7129038279999804,
                    "count": 1,
                    "self": 0.7129038279999804
                },
                "TrainerController.advance": {
                    "total": 11.646286337000447,
                    "count": 203,
                    "self": 0.002890762003062264,
                    "children": {
                        "env_step": {
                            "total": 11.643395574997385,
                            "count": 203,
                            "self": 8.072210629995652,
                            "children": {
                                "SubprocessEnvManager._take_step": {
                                    "total": 3.567691051001475,
                                    "count": 203,
                                    "self": 0.020044783000457755,
                                    "children": {
                                        "TorchPolicy.evaluate": {
                                            "total": 3.5476462680010172,
                                            "count": 203,
                                            "self": 0.06425901000329759,
                                            "children": {
                                                "TorchPolicy.sample_actions": {
                                                    "total": 3.4833872579977196,
                                                    "count": 203,
                                                    "self": 3.4833872579977196
                                                }
                                            }
                                        }
                                    }
                                },
                                "workers": {
                                    "total": 0.003493894000257569,
                                    "count": 202,
                                    "self": 0.0,
                                    "children": {
                                        "worker_root": {
                                            "total": 12.320292559997824,
                                            "count": 202,
                                            "is_parallel": true,
                                            "self": 4.885522642997785,
                                            "children": {
                                                "run_training.setup": {
                                                    "total": 0.0,
                                                    "count": 0,
                                                    "is_parallel": true,
                                                    "self": 0.0,
                                                    "children": {
                                                        "steps_from_proto": {
                                                            "total": 0.011005758000010246,
                                                            "count": 1,
                                                            "is_parallel": true,
                                                            "self": 0.0003166559999954188,
                                                            "children": {
                                                                "_process_visual_observation": {
                                                                    "total": 0.010426326000015251,
                                                                    "count": 2,
                                                                    "is_parallel": true,
                                                                    "self": 0.00026279399958184513,
                                                                    "children": {
                                                                        "observation_to_np_array": {
                                                                            "total": 0.010163532000433406,
                                                                            "count": 6,
                                                                            "is_parallel": true,
                                                                            "self": 0.00013330100000530365,
                                                                            "children": {
                                                                                "process_pixels": {
                                                                                    "total": 0.010030231000428103,
                                                                                    "count": 6,
                                                                                    "is_parallel": true,
                                                                                    "self": 0.0007816070005901565,
                                                                                    "children": {
                                                                                        "image_decompress": {
                                                                                            "total": 0.009248623999837946,
                                                                                            "count": 6,
                                                                                            "is_parallel": true,
                                                                                            "self": 0.009248623999837946
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        }
                                                                    }
                                                                },
                                                                "_process_vector_observation": {
                                                                    "total": 0.000262775999999576,
                                                                    "count": 2,
                                                                    "is_parallel": true,
                                                                    "self": 0.000262775999999576
                                                                }
                                                            }
                                                        },
                                                        "UnityEnvironment.step": {
                                                            "total": 0.08557706200008397,
                                                            "count": 1,
                                                            "is_parallel": true,
                                                            "self": 0.0006010500003412744,
                                                            "children": {
                                                                "UnityEnvironment._generate_step_input": {
                                                                    "total": 0.00016605299992988876,
                                                                    "count": 1,
                                                                    "is_parallel": true,
                                                                    "self": 0.00016605299992988876
                                                                },
                                                                "communicator.exchange": {
                                                                    "total": 0.08177106199991613,
                                                                    "count": 1,
                                                                    "is_parallel": true,
                                                                    "self": 0.08177106199991613
                                                                },
                                                                "steps_from_proto": {
                                                                    "total": 0.0030388969998966786,
                                                                    "count": 1,
                                                                    "is_parallel": true,
                                                                    "self": 0.0002606250000098953,
                                                                    "children": {
                                                                        "_process_visual_observation": {
                                                                            "total": 0.002563538000003973,
                                                                            "count": 2,
                                                                            "is_parallel": true,
                                                                            "self": 0.00036613600013879477,
                                                                            "children": {
                                                                                "observation_to_np_array": {
                                                                                    "total": 0.0021974019998651784,
                                                                                    "count": 6,
                                                                                    "is_parallel": true,
                                                                                    "self": 0.00013559099988924572,
                                                                                    "children": {
                                                                                        "process_pixels": {
                                                                                            "total": 0.0020618109999759326,
                                                                                            "count": 6,
                                                                                            "is_parallel": true,
                                                                                            "self": 0.0009174289998554741,
                                                                                            "children": {
                                                                                                "image_decompress": {
                                                                                                    "total": 0.0011443820001204585,
                                                                                                    "count": 6,
                                                                                                    "is_parallel": true,
                                                                                                    "self": 0.0011443820001204585
                                                                                                }
                                                                                            }
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        },
                                                                        "_process_vector_observation": {
                                                                            "total": 0.00021473399988281017,
                                                                            "count": 2,
                                                                            "is_parallel": true,
                                                                            "self": 0.00021473399988281017
                                                                        }
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    }
                                                },
                                                "UnityEnvironment.step": {
                                                    "total": 7.43476991700004,
                                                    "count": 201,
                                                    "is_parallel": true,
                                                    "self": 0.09699524199936604,
                                                    "children": {
                                                        "UnityEnvironment._generate_step_input": {
                                                            "total": 0.03602335800133005,
                                                            "count": 201,
                                                            "is_parallel": true,
                                                            "self": 0.03602335800133005
                                                        },
                                                        "communicator.exchange": {
                                                            "total": 6.89298212099834,
                                                            "count": 201,
                                                            "is_parallel": true,
                                                            "self": 6.89298212099834
                                                        },
                                                        "steps_from_proto": {
                                                            "total": 0.40876919600100337,
                                                            "count": 201,
                                                            "is_parallel": true,
                                                            "self": 0.03866307800353752,
                                                            "children": {
                                                                "_process_visual_observation": {
                                                                    "total": 0.3270108069980324,
                                                                    "count": 402,
                                                                    "is_parallel": true,
                                                                    "self": 0.027272977997427006,
                                                                    "children": {
                                                                        "observation_to_np_array": {
                                                                            "total": 0.2997378290006054,
                                                                            "count": 1206,
                                                                            "is_parallel": true,
                                                                            "self": 0.02095674199631503,
                                                                            "children": {
                                                                                "process_pixels": {
                                                                                    "total": 0.2787810870042904,
                                                                                    "count": 1206,
                                                                                    "is_parallel": true,
                                                                                    "self": 0.104499624999562,
                                                                                    "children": {
                                                                                        "image_decompress": {
                                                                                            "total": 0.17428146200472838,
                                                                                            "count": 1206,
                                                                                            "is_parallel": true,
                                                                                            "self": 0.17428146200472838
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        }
                                                                    }
                                                                },
                                                                "_process_vector_observation": {
                                                                    "total": 0.04309531099943342,
                                                                    "count": 402,
                                                                    "is_parallel": true,
                                                                    "self": 0.04309531099943342
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                },
                "trainer_threads": {
                    "total": 7.650099996681092e-05,
                    "count": 1,
                    "self": 7.650099996681092e-05,
                    "children": {
                        "thread_root": {
                            "total": 0.0,
                            "count": 0,
                            "is_parallel": true,
                            "self": 0.0,
                            "children": {
                                "trainer_advance": {
                                    "total": 11.527443817012227,
                                    "count": 19114,
                                    "is_parallel": true,
                                    "self": 0.2529891150329604,
                                    "children": {
                                        "process_trajectory": {
                                            "total": 11.274454701979266,
                                            "count": 19114,
                                            "is_parallel": true,
                                            "self": 11.274454701979266
                                        }
                                    }
                                }
                            }
                        }
                    }
                },
                "TrainerController._save_models": {
                    "total": 0.16066428699991775,
                    "count": 1,
                    "self": 0.0039965840001059405,
                    "children": {
                        "RLTrainer._checkpoint": {
                            "total": 0.1566677029998118,
                            "count": 1,
                            "self": 0.1566677029998118
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by awjuliani** [comment]
Can you share some about the environment you are training? What kinds of observations are there? The slow-down is likely due to the environment, as you say.

**Comment by simon-winter** [comment]
First, thank you for your help!

I inspected the environment closer and can describe as follows:

a few simple bool, Vector 3 observations, about 10. the output is just like tank controls l/r wheel forward/stop/back, so super simple.
Additionally though i want to provide an map for the agent of its surroundings. a very simple pixelraster (30x30) where 3 diffrent colors stand for obstacle/floor/not set. A data based minimap so to speak. This raster i want to put through the CNN by providing the pxielraster as visual observation, as that seems much more reasonable then putting the pixelraster in as 900 individual int observations. 
Using RenderTexture or camera over texture just seems so much overhead, so i use the provided custom grid observation, which does visual observations on my 2d int-array-pixelgrid.
Also i use the SAC trainer.

When starting the training i checked 1k step summarys and had a go: 
- the first 10k steps go really quick (10 seconds between sumamries). This 10k steps corrospond to my setting of `buffer_init_steps: 10000`
- suddendly it takes 3 minutes for 1k steps
- next it takes 20 minutes to train 1ksteps

so a rapid descent in progress.
![grafik](https://user-images.githubusercontent.com/34577718/101817827-4310ca00-3b23-11eb-9b2a-19b1028b38ba.png)

so is it possible that after the `buffer_init_steps` the pixelgrid data starts accumulating in the experience buffer and that just keeps bloating up the amount of data? (Which unit is the `buffer_size: 50000` in? 50K experiences so it scales with map? or is it fixed like byte? is the experience buffer variable in size and not prefilled with my pixelgrid-fake data? If so it shouldn't grow difficulter over time, right?

Which process in pytorch takes longer over time and why? shouldn't training difficultie stay constant?

Also i did check my envirnment for memory leaks ect!
 - taskmanager shows cosntant memory usage (~60%)
- cpu also sits at 60% )why can't i get 100% consistently while training? server never went above 60% with 8 cores x3.6ghz)
- i also checked that i free unity specific memory things by calling Resources.UnloadUnusedAssets(); every episode begin + not allocating unity objects/components like `var bla = new Texture2D` in the first place. (only allocating basic c# types with new)
-same goes for arrays and lists , i dont reallocate even.
the scene only uses very simple wheel physic collider and a little basic vector math. nothing taxing in update loops scene runs at several 100fps
the only not very basic operation i do is raycasting a bit, but thats literly the only thing that actually happens besides minor physics simulation (its a cleaning robot simulation and i want to train the cleaning robot). alos its only a handfull of raycasts.

Note: i am currently training wihtout CNN and use the 1000 basic int observations. its quite slow, but it stays reasonable quick for atleast 50k steps and is still going. (i'll update here incase it also gets stuck/decreases in speed but i don't think that will happen)
![grafik](https://user-images.githubusercontent.com/34577718/101818955-e6161380-3b24-11eb-871e-3b139c7bbd47.png)

**Comment by simon-winter** [comment]
![grafik](https://user-images.githubusercontent.com/34577718/101819172-25dcfb00-3b25-11eb-98f7-5c6c36a14ba4.png)
my trainign file

**Comment by awjuliani** [comment]
Hi @simon-winter 

Thank you for providing all of this additional information. Due to how SAC and PPO work you will see different behaviors. In general, SAC will be slower, as it performs more training per episode step of the environment. What may be happening is that as the buffer gets larger, it eventually passes the minimum size to start sampling from it for network updates (10000 items). I would experiment with much smaller values for the buffer and batch and share the results. Are the actions discrete or continuous? If they are discrete, you can likely get away with an order of magnitude smaller values for these buffer parameters.

**Comment by simon-winter** [comment]
i have choosen the largest "typical range" provided in the wiki, as my robot shall clean the entire room with strategy. so i thoughed i need him to collect alot of experience to find a good startegy for a longterm activity.
I'll try siginifantly smaller batch and puffer and call results back, 
the ppo training seems also to get steadily slower, just at a smaller haste, so that would make sense to be the same reason.
But what exactly you mena by the network sampling? pytorch can't keep up with training on so much data which i send all the time? or does unity stop sending data if its too much?

**Comment by awjuliani** [comment]
In PPO the updating of the neural network happens when the buffer is filled up, and then the buffer is emptied. In SAC, once the buffer is filled (enough), then updating the network happens more or less every episode step, leading to a greater slow-down in terms of number of episodes.

**Comment by simon-winter** [comment]
ok i still dont get the concept. The buffer gets filled with our observations data. once the buffer is full it will get send to python and that trains with the data. wouldn't smaller buffer sizes trigger the sending process to start just much quicker? if i have husge buffers i can collect the data and the network trains as quickly as it can and i feed it liek a factory with data all the time. why is it faster to train if that buffer is much smaller?

But thats defintly it! Thank you so much!
![grafik](https://user-images.githubusercontent.com/34577718/101824454-815eb700-3b2c-11eb-8e2a-4b28d95824bb.png)

any downsides when i pick a buffer too small?

**Comment by simon-winter** [comment]
ah i think i got it now!

more data means more time to train but better representation of whats happening. If thats too big we train long and this will always scale from this initial difficulty!
picking a smaller data amount to trian, makes it less "representing" of the scene, but pytorch can train quick enough from it, that the scaling in upcoming training doesnt matter so much

**Comment by simon-winter** [comment]
But it seemed to encounter a diffrent problem: the model is not exported as the last modified suggests. is that aborted message related to a specif failure?

![grafik](https://user-images.githubusercontent.com/34577718/101828907-7d359800-3b32-11eb-8b7c-7d072294894d.png)

**Comment by awjuliani** [comment]
According to that log, it looks like the ONNX file was indeed saved. Does it not appear in the listed location?

**Comment by simon-winter** [comment]
im force overwriting the same test set, so  i rely on the last modified time provided by winscp. i was just bugged by the Aborted (core dumped) messageoccuring with that not written .onnx. maybe not saving it to that .onnx was just a onetime problem, i'm in the process of traing again and i now let it run uninterrupted. i'll see tomorrow if saving worked or if its reproducable. 
Thank you for your help!

**Comment by simon-winter** [comment]
nope, it does indeed not save that mdoel. the specified path doesnt exist.that Aborted abdendum is telling the truth

**Comment by simon-winter** [comment]
i just interrupted it, quick interuupts will write the file, but if i let it run for 10 minutes it states aborted and the file is not written

**Comment by simon-winter** [comment]
oh man, nvm... its late... i restarted WinSCP and now the file is shown correctly... so THank you very much, this fixes literly all my problems. Thank you!
